## Final Project Submission

Please fill out:
* Student name: 
* Student pace: self paced / part time / full time
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:


In [1]:
import pandas as pd
import numpy as np

from dateutil.relativedelta import relativedelta
from math import radians, cos, sin, asin, sqrt

# Exploration

In [2]:
data = pd.read_csv('kc_house_data.csv')
data.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0         NaN   0.0  ...      7        1180            0.0   
1      7242     2.0         0.0   0.0  ...      7        2170          400.0   
2     10000     1.0         0.0   0.0  ...      6         770            0.0   
3      5000     1.0         0.0   0.0  ...      7        1050          910.0   
4      8080     1.0         0.0   0.0  ...      8        1680            0.0   

  yr_built  yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  
0     1955           0.0    98178  47.5112 -122.257           1340        5650  
1     1951        1991.0    98125  47.7210 -122.319           1690        7639  
2     1933           NaN    98028  47.7379 -122.233           2720        8062  
3     1965           0.0    98136  47.5208 -122.393           1360        5000  
4     1987           0.0    98074  47.6168 -122.045           1800        7503  

[5 rows x 21 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
id               21597 non-null int64
date             21597 non-null object
price            21597 non-null float64
bedrooms         21597 non-null int64
bathrooms        21597 non-null float64
sqft_living      21597 non-null int64
sqft_lot         21597 non-null int64
floors           21597 non-null float64
waterfront       19221 non-null float64
view             21534 non-null float64
condition        21597 non-null int64
grade            21597 non-null int64
sqft_above       21597 non-null int64
sqft_basement    21597 non-null object
yr_built         21597 non-null int64
yr_renovated     17755 non-null float64
zipcode          21597 non-null int64
lat              21597 non-null float64
long             21597 non-null float64
sqft_living15    21597 non-null int64
sqft_lot15       21597 non-null int64
dtypes: float64(8), int64(11), object(2)
memory usage: 3.5+ MB


In [4]:
data['id'].nunique()

21420

In [5]:
data.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.159700e+04  21597.000000  19221.000000  21534.000000  21597.000000   
mean   1.509941e+04      1.494096      0.007596      0.233863      3.409825   
std    4.141264e+04      0.539683      0.086825      0.765686      0.650546   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068500e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above      yr_built  yr_renovated       zipcode  \
count  21597.000000  21597.000000  21597.000000  17755.000000  21597.000000   
mean       7.657915   1788.596842   1970.999676     83.636778  98077.951845   
std        1.173200    827.759761     29.375234    399.946414     53.513072   
min        3.000000    370.000000   1900.000000      0.000000  98001.000000   
25%        7.000000   1190.000000   1951.000000      0.000000  98033.000000   
50%        7.000000   1560.000000   1975.000000      0.000000  98065.000000   
75%        8.000000   2210.000000   1997.000000      0.000000  98118.000000   
max       13.000000   9410.000000   2015.000000   2015.000000  98199.000000   

                lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000   21597.000000   21597.000000  
mean      47.560093   -122.213982    1986.620318   12758.283512  
std        0.138552      0.140724     685.230472   27274.441950  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471100   -122.328000    1490.000000    5100.000000  
50%       47.571800   -122.231000    1840.000000    7620.000000  
75%       47.678000   -122.125000    2360.000000   10083.000000  
max       47.777600   -121.315000    6210.000000  871200.000000

# Cleaning

### Convert datatypes

In [6]:
#Convert dates and years to datetime.
data['date']= pd.to_datetime(data['date'])
data['yr_built']= pd.to_datetime(data['yr_built'], 
                                 format = '%Y')

In [7]:
#Trying to convert sqft_basement to float called an error.
data['sqft_basement'].value_counts()

0.0       12826
?           454
600.0       217
500.0       209
700.0       208
          ...  
602.0         1
1770.0        1
1798.0        1
516.0         1
2730.0        1
Name: sqft_basement, Length: 304, dtype: int64

The ? in the sqft_basement column won't work. I'll use the other data I have to deduce the values for this column. 

In [8]:
data['sqft_basement']=(data['sqft_living']-data['sqft_above'])
data['sqft_basement'].describe()

count    21597.000000
mean       291.725008
std        442.667800
min          0.000000
25%          0.000000
50%          0.000000
75%        560.000000
max       4820.000000
Name: sqft_basement, dtype: float64

Zipcode is not a continuous variable. We will need to use one-hot encoding to make it into a useable categorical variable.

In [9]:
data['zipcode'].nunique()

70

In [10]:
data['zipcode'].astype('category', copy=False)
zipdummies = pd.get_dummies(data['zipcode'])
data = data.join(zipdummies, how='outer')
data.drop('zipcode', axis=1, inplace=True)

### Address Obvious Outliers

In [11]:
data[data['bedrooms']>30]

id       date     price  bedrooms  bathrooms  sqft_living  \
15856  2402100895 2014-06-25  640000.0        33       1.75         1620   

       sqft_lot  floors  waterfront  view  ...  98146  98148  98155  98166  \
15856      6000     1.0         0.0   0.0  ...      0      0      0      0   

      98168  98177  98178  98188  98198  98199  
15856     0      0      0      0      0      0  

[1 rows x 90 columns]

33 bedrooms in a 1620 sqft house seems like it was likely a typo. I'll impute 3 as the number of rooms.  

In [12]:
data.loc[data['bedrooms']>30, 'bedrooms']=3

### Duplicates

In [13]:
data.duplicated().sum()

0

### Null Values

In [14]:
data.isna().sum()

id           0
date         0
price        0
bedrooms     0
bathrooms    0
            ..
98177        0
98178        0
98188        0
98198        0
98199        0
Length: 90, dtype: int64

In [15]:
display(data['waterfront'].value_counts())
display(data['view'].value_counts())
display(data['yr_renovated'].value_counts())

0.0    19075
1.0      146
Name: waterfront, dtype: int64

0.0    19422
2.0      957
3.0      508
1.0      330
4.0      317
Name: view, dtype: int64

0.0       17011
2014.0       73
2003.0       31
2013.0       31
2007.0       30
          ...  
1946.0        1
1959.0        1
1971.0        1
1951.0        1
1954.0        1
Name: yr_renovated, Length: 70, dtype: int64

For these particular variables, which are skewed so heavily toward zero, I will simply impute 0.

In [16]:
data.loc[data['waterfront'].isna(), 'waterfront']=0
data.loc[data['view'].isna(), 'view']=0
data.loc[data['yr_renovated'].isna(), 'yr_renovated']=0

Now that the null values have been dealt with I can properly convert 'waterfront' to categorical datatype.  

In [17]:
data['waterfront'].astype('category', copy=False)
wfdummies = pd.get_dummies(data['waterfront'])
data = data.join(wfdummies, how='outer')
data.drop('waterfront', axis=1, inplace=True)

# Feature Creation

I want to be able to use age to describe the houses, so I'll create a feature that stores their age at sale. 

### Age

In [18]:
data['age'] = [relativedelta(data['date'][i], 
                             data['yr_built'][i]).years 
               for i in range(0,len(data['date']))]
data['age'].describe()

count    21597.000000
mean        43.323841
std         29.376456
min          0.000000
25%         18.000000
50%         40.000000
75%         63.000000
max        115.000000
Name: age, dtype: float64

### City Data
A more decipherable way to utilize the latitude and longitude data is to calculate the distance between each house and the cities in King County. I can use that information to estimate what city that house resides in, providing another means for analysis. 

I'll use the haversine formula to calculate the miles of distance from the center of particular cities to each house. I'll then use that formula to create columns for the distance to each city and additional columns for the categorical information of what city each house resides in. For this analysis I'll use the largest 6 cities by population in the King County area, including Tacoma, which is technically in Pierce county. 

In [19]:
def haversine(city):
    """
    Return distance in miles between house and city.
    
    Parameters
    city-- tuple, coordinates(lon,lat) for a city
    
    """
    dists = []
    for i in range(0,len(data['long'])):
        lon1, lat1, lon2, lat2 = map(radians, 
                                     [data['long'][i],
                                      data['lat'][i],
                                      city[0], city[1]])
        dlon = lon2 - lon1 
        dlat = lat2 - lat1 
        a = (sin(dlat/2)**2 + cos(lat1) 
             * cos(lat2) * sin(dlon/2)**2)
        c = 2 * asin(sqrt(a)) 
        r = 3956 # Radius of earth in miles. 
        dists.append(round(c*r,2))
    return np.array(dists)       

In [20]:
Seattle = (-122.335167, 47.608013, 'Seattle')
haversine(Seattle)

array([ 7.61,  7.84, 10.15, ...,  1.93, 13.4 ,  1.94])

In [21]:
Seattle = (-122.335167, 47.608013, 'Seattle')
Bellevue = (-122.1994444, 47.6105556, 'Bellevue')
Renton = (-122.2158333, 47.4830556, 'Renton') 
Kent = (-122.2336111, 47.3811111, 'Kent')
Federal_Way = (-122.3113889, 47.3225, 'Federal Way')
Tacoma = (-122.4430556, 47.2530556, 'Tacoma')

city_list = [Seattle, Bellevue, Renton, 
             Kent, Federal_Way, Tacoma]

for city in city_list:
    data['dist_to_{}'.format(city[2])] = haversine(city)

In [22]:
city_dist_cols = [('dist_to_{}'.format(city[2]), city[2]) 
                  for city in city_list]
nearest_cities = []

for i in range(0,len(data['long'])):
    dists = []
    for city_dist in city_dist_cols:
        dists.append(((data[city_dist[0]][i]), city_dist[1]))
    nearest_cities.append(sorted(dists)[0][1])
    
    
data['nearest_city'] = nearest_cities

In [23]:
data[['dist_to_Seattle', 'dist_to_Bellevue',
     'dist_to_Renton', 'dist_to_Kent',
     'dist_to_Federal Way', 'dist_to_Tacoma',
     'nearest_city']]

dist_to_Seattle  dist_to_Bellevue  dist_to_Renton  dist_to_Kent  \
0                 7.61              7.37            2.73          9.05   
1                 7.84              9.44           17.12         23.80   
2                10.15              8.93           17.61         24.63   
3                 6.60             10.94            8.67         12.18   
4                13.52              7.20           12.19         18.50   
...                ...               ...             ...           ...   
21592             6.32              9.17           16.11         22.59   
21593             6.83             10.24            7.08         10.77   
21594             1.93              4.77            8.61         15.04   
21595            13.40              8.03            7.71         13.09   
21596             1.94              4.77            8.59         15.02   

       dist_to_Federal Way  dist_to_Tacoma nearest_city  
0                    13.27           19.83       Renton  
1                    27.52           32.82      Seattle  
2                    28.91           34.88     Bellevue  
3                    14.21           18.63      Seattle  
4                    23.82           31.25     Bellevue  
...                    ...             ...          ...  
21592                26.07           31.14      Seattle  
21593                13.21           18.19      Seattle  
21594                18.78           24.51      Seattle  
21595                18.51           26.14       Renton  
21596                18.76           24.49      Seattle  

[21597 rows x 7 columns]

In [24]:
data['nearest_city'].value_counts()

Seattle        7210
Bellevue       6482
Renton         3582
Kent           2747
Federal Way    1562
Tacoma           14
Name: nearest_city, dtype: int64

In [25]:
data['nearest_city'].astype('category', copy=False)
ncdummies = pd.get_dummies(data['nearest_city'])
data = data.join(ncdummies, how='outer')
data.drop('nearest_city', axis=1, inplace=True)

### Size Relative to Neighbors

I'd like to test the old real estate addage that the biggest house on the block has lower than expected value by creating an additional feature that tracks the size of a home compared with its nearst 15 neighbors.

In [28]:
rel_size_living = data['sqft_living']/data['sqft_living15']
rel_size_lot = data['sqft_lot']/data['sqft_lot15']

# Build a Baseline Model